### ISSUE TITLE
Should foreign language wrappers aware about array_proxy?

### ISSUE THREAD

**Issue by unbornchikken** [arrayfire/arrayfire#856]
Hi,

While implementing [Node.js wrapper](https://github.com/arrayfire/arrayfire_js) I've realised that some operations like indexing returns array_proxy instead of array. At the current implementation I don't care about this class, just use its implicit array conversion operator and store the array instance in my wrapper class: https://github.com/arrayfire/arrayfire_js/blob/master/src/arraywrapper.h#L126 So in Node.js side the mentioned operations like indexing involves this array conversion also. I've noticed that this conversion involves some mumbo-jumbo, doesn't seem too O(1) to mee. The question is, should my wrapper aware of this, and should there be an array_proxy prointer too? What's the purpose of this array_proxy class anyway?

Thanks.

**Comment by umar456** [comment]
TLDR
--------
array_proxy is an implementation of the proxy pattern which assists in assignment and indexing. Anyone using ArrayFire should not be using/creating this object directly. For all intents and purposes, it should be treated as an af::array object.

Proxy Class
----------------
The `array_proxy` class is created when performing indexing and assignment. It is returned by indexing operations like `operator()`, `row(s)/col(s)` etc. and its purpose it to figure out wether the indexing is being performed on the left hand side(assignment) or the right hand side(indexing/referencing). For example:

~~~cpp
array a = randu(10, 10);
array b = randu(5);

a(seq(5), 1) = b;
~~~

Here you are assigning the `b` vector to the first 5 values in the second column of `a`. The `a` index will return an `array_proxy` class. The `operator=` is called object which in turn calls `af_assign_gen` in the C API. If on the other hand `a` was being assigned to `b`: 

~~~cpp
array a = randu(10, 10);
array b = randu(5);

b = a(seq(5), 1);
~~~

The `operator()` function will again return an `array_proxy` class but this time the `operator array`(implicit conversion operator) will be called because it is on the right hand side. This function in turn calls the `af_index_gen` C call and the data will therefore be referenced.

There are other ways to do this but it is less efficient and adds unnecessary bloat to the `af::array` object.

More Info:
------------------
https://en.wikibooks.org/wiki/More_C%2B%2B_Idioms/Temporary_Proxy

Language Integration
---------------------------
Ideally language integrators would make use of the C API because of compiler name mangling issues in C++. In fact out C++ API is built on top of the C API because of name mangling issues. This is not always straightforward because the C API is not as easy to use as the C++ API. Each language will have its own mechanism to handle indexing so this technique might not be useful everywhere.

**Comment by unbornchikken** [comment]
Ok, so I'm good with that array poiner then. Just for the record since v8 is in C++ and Node wrapper is just a v8 wrapper, there is no reason to reinvent the C++ API again, I'm happy with the included one. ;)

**Comment by unbornchikken** [comment]
Just for the other record, when I started the work, I had considered to use the C API with node-ffi without a C++ layer whatsoverer, but unfortunatelly ffi libraries are too slow to wrap anything that is about the performance. There are too much overhead involved on each call.

**Comment by umar456** [comment]
Yeah, for this project were you are making use of the V8 engine, so the C++ interface is fine. I was referring to other languages were name mangling would be an issue.

**Comment by unbornchikken** [comment]
Yeah, ive realised it since.Thanks for the explanation, i think i should review my assign method because of this.
